src/shallow/ShallowWrapper.lua

```diff
@@ -0,0 +1,232 @@
+		for _, subVirtualNode in pairs(virtualNode.children) do
+			getChildren(subVirtualNode, results, maxDepth)
+		end
+	else
+		local childWrapper = ShallowWrapper.new(
+			virtualNode,
+			maxDepth
+		)
+
+		table.insert(results, childWrapper)
+	end
+end
+
+local function filterProps(props)
+	if props[Children] == nil then
+		return props
+	end
+
+	local filteredProps = {}
+
+	for key, value in pairs(props) do
+		if key ~= Children then
+			filteredProps[key] = value
+		end
+	end
+
+	return filteredProps
+end
+
+function ShallowWrapper.new(virtualNode, maxDepth)
+	virtualNode = findNextVirtualNode(virtualNode, maxDepth)
+
+	local wrapper = {
+		_virtualNode = virtualNode,
+		_childrenMaxDepth = maxDepth - 1,
+		_virtualNodeChildren = maxDepth == 0 and {} or virtualNode.children,
+		type = getTypeFromVirtualNode(virtualNode),
+		props = filterProps(virtualNode.currentElement.props),
+		hostKey = virtualNode.hostKey,
+		instance = virtualNode.hostObject,
+	}
+
+	return setmetatable(wrapper, ShallowWrapperMetatable)
+end
+
+function ShallowWrapper:childrenCount()
+	local count = 0
+
+	for _, virtualNode in pairs(self._virtualNodeChildren) do
+		local element = virtualNode.currentElement
+		count = count + countChildrenOfElement(element)
+	end
+
+	return count
+end
+
+function ShallowWrapper:find(constraints)
+	for constraint in pairs(constraints)	 do
+		if not ContraintFunctions[constraint] then
+			error(('unknown constraint %q'):format(constraint))
+		end
+	end
+
+	local results = {}
+	local children = self:getChildren()
+
+	for i=1, #children do
+		local childWrapper = children[i]
+		if childWrapper:_satisfiesAllContraints(constraints) then
+			table.insert(results, childWrapper)
+		end
+	end
+
+	return results
+end
+
+function ShallowWrapper:findUnique(constraints)
+	local children = self:getChildren()
+
+	if constraints == nil then
+		assert(
+			#children == 1,
+			("expect to contain exactly one child, but found %d"):format(#children)
+		)
+		return children[1]
+	end
+
+	local constrainedChildren = self:find(constraints)
+
+	assert(
+		#constrainedChildren == 1,
+		("expect to find only one child, but found %d"):format(#constrainedChildren)
+	)
+
+	return constrainedChildren[1]
+end
+
+function ShallowWrapper:getChildren()
+	local results = {}
+
+	for _, childVirtualNode in pairs(self._virtualNodeChildren) do
+		getChildren(childVirtualNode, results, self._childrenMaxDepth)
+	end
+
+	return results
+end
+
+function ShallowWrapper:matchSnapshot(identifier)
+	assert(typeof(identifier) == "string", "Snapshot identifier must be a string")
+
+	local snapshotResult = snapshot(identifier, self)
+
+	snapshotResult:match()
+end
+
+function ShallowWrapper:getSnapshotString(identifier)
+	assert(typeof(identifier) == "string", "Snapshot identifier must be a string")
+
+	local snapshotResult = snapshot(identifier, self)
+
+	return snapshotResult:getSnapshotString()
+end
+
+function ShallowWrapper:_satisfiesAllContraints(constraints)
```

Comment:
this is a good candidate to not be a member function to make it truly private. Consumers will get access to ShallowWrapper directly, so we should hide as much as we can versus using the underscore naming convention

Reply:
It could also make sense to put all these utility functions into their own file so it's super clear that everything in ShallowWrapper is available to consumers